ndb/tests/unit/test_model.py

```diff
@@ -2283,16 +2295,163 @@ def test_constructor():
+
+            class Mine(model.Model):
+                @classmethod
+                def _get_kind(cls):
+                    return 525600
+
+    @staticmethod
+    def test_invalid_property_name():
+        with pytest.raises(TypeError):
+
+            class Mine(model.Model):
+                _foo = model.StringProperty()
+
+    @staticmethod
+    def test_repeated_property():
+        class Mine(model.Model):
+            foo = model.StringProperty(repeated=True)
+
+        assert Mine._has_repeated
+
+    @staticmethod
+    def test_non_property_attribute():
+        model_attr = unittest.mock.Mock(spec=model.ModelAttribute)
+
+        class Mine(model.Model):
+            baz = model_attr
+
+        model_attr._fix_up.assert_called_once_with(Mine, "baz")
 
 
 class TestModel:
     @staticmethod
-    def test_constructor():
-        with pytest.raises(NotImplementedError):
-            model.Model()
+    def test_constructor_defaults():
+        entity = model.Model()
+        assert entity.__dict__ == {"_values": {}}
+
+    @staticmethod
+    def test_constructor_key():
+        key = key_module.Key("Foo", "bar")
+        entity = model.Model(key=key)
+        assert entity.__dict__ == {"_values": {}, "_entity_key": key}
+
+        entity = model.Model(_key=key)
+        assert entity.__dict__ == {"_values": {}, "_entity_key": key}
+
+    @staticmethod
+    def test_constructor_key_parts():
+        entity = model.Model(id=124)
+        key = key_module.Key("Model", 124)
+        assert entity.__dict__ == {"_values": {}, "_entity_key": key}
+
+    @staticmethod
+    def test_constructor_key_and_key_parts():
+        key = key_module.Key("Foo", "bar")
+        with pytest.raises(exceptions.BadArgumentError):
+            model.Model(key=key, id=124)
+
+    @staticmethod
+    def test_constructor_user_property_collision():
+        class SecretMap(model.Model):
+            key = model.IntegerProperty()
+
+        entity = SecretMap(key=1001)
+        assert entity.__dict__ == {"_values": {"key": 1001}}
+
+    @staticmethod
+    def test_constructor_with_projection():
+        class Book(model.Model):
+            pages = model.IntegerProperty()
+            author = model.StringProperty()
+            publisher = model.StringProperty()
+
+        entity = Book(
+            pages=287, author="Tim Robert", projection=("pages", "author")
+        )
+        assert entity.__dict__ == {
+            "_values": {"pages": 287, "author": "Tim Robert"},
+            "_projection": ("pages", "author"),
+        }
+
+    @staticmethod
+    def test_constructor_non_existent_property():
+        with pytest.raises(AttributeError):
+            model.Model(pages=287)
+
+    @staticmethod
+    def test_constructor_non_property():
+        class TimeTravelVehicle(model.Model):
+            speed = 88
+
+        with pytest.raises(TypeError):
+            TimeTravelVehicle(speed=28)
+
+    @staticmethod
+    def test_repr():
```

Comment:
@chrisrossi I don't like that I jammed 5 actual tests into a single test, but I did it somewhat to avoid having to redefine `ManyFields` in each test. Do you think I should split this up?

Reply:
I would split it up. I tend to do a lot of copy/paste for tests.  I guess you could also declare ManyFields at module scope, if you wanted to avoid repetition.

Reply:
Done.